rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,161 @@
+---
+authors: Michael Wilson (michael.wilson@goteleport.com)
+state: draft
+---
+
+# RFD 124 - Plugin static credentials
+
+### Required Approvers
+
+* Engineering @r0mant, @justinas
+* Security @reed
+* Product: @klizhentas
+
+## What
+
+Allow storage of plugin static credentials within Teleport in a safe and secure way.
+
+## Why
+
+As we grow our plugin support, we have discovered that the various services we
+hope to integrate do not offer a uniform mechanism for configuration. For instance,
+Slack uses oauth2, but Okta requires an API token, and Jamf requires a username and
+password.
+
+Today, the plugin houses a credentials object that allows for this sort of storage.
+However, unlike an oauth authorization code, static credentials pose a much greater risk of
+exposure, and rotating credentials can be a time consuming process.  To reduce the risks associated
+with storing static credentials within Teleport and to allow for easier credentials rotation,
+we would like to create a new resource within Teleport that is handled in an explicitly secure
+way.
+
+## Details
+
+### The `PluginStaticCredentials` object
+
+A new object will be needed to house static credentials. At present it will support:
+
+- API tokens
+- Usernames and passwords
+- OAuth client ID and secret
+
+This object has very strict access control requirements that will be explained in detail
+in the **Access Control** section.
+
+```yaml
+kind: plugin_static_credentials
+version: v1
+metadata:
+  name: plugin-name # name must match the name of the plugin.
+spec:
+  last_rotated: "2023-05-09T16:50:52.497874Z"
+  credentials:
+    # Only one of these credential types must be defined at a time. All values in this
+    # are string literals and not file locations.
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+    oauth_client_secret:
+      client_id: example-client-id
+      client_secret: example-client-secret
+```
+
+If a plugin requires static credentials, there is a 1-to-1 mapping between `Plugin`
+objects and `PluginStaticCredentials` objects. If a `Plugin` is using oauth, however,
+it will not need a `PluginStaticCredentials` and none will be created. The object will
+inherit the same name as its associated plugin.
+
+### Changes to the `Plugin` object
+
+The `Plugin` object's `Credentials` field will have a new type of credential called
+`PluginNeedsStaticCredentials` with a single boolean `Enabled` flag. If this flag is set,
```

Comment:
I think we can have an empty `message PluginNeedsStaticCredentials {}`, i.e. proto can hold a `oneof` variant even if it has no fields set (not 100% sure though). Does the `Enabled` flag have a real use if so?

---

I would suggest a different name such as `message PluginStaticCredentialsRef[erence]`. Perhaps in the future we may even support several instances of a plugin to share credentials, and expand the variant to something like:

```proto
message PluginStaticCredentialsRef {
  string name = 1; // when empty, defaults to the old behavior (same as plugin name) 
}
```

Reply:
I was originally thinking just having it's presence as well but I convinced myself that there was a worry with recognizing `Empty but present` in the underlying JSON. 

Sounds good on the rename.

Reply:
I've adjusted this a bit so that it's now got some semantically interesting fields, so let me know what you think.